### DISCUSSION TITLE
Planning for a volume testing | Is there any APNs restriction

### DISCUSSION THREAD

**Discussion by inventriz** [relayrides/pushy#978] [Q&A] [answered]
Hi Team,

We are doing a quick POC to use pushy in our project to send push notification to our app. We would like to do a basic volume testing to check if there is any drops, end-to-end response time etc.
We have limited number of test devices to do this testing. So eventually we need to generate load using one or two device token(s) only.  

With this context does APNs have any restriction if it observes load coming from one single source and going to one/two device tokens? Will there be any chance that our app may get blocked (or something) or IPs are blocked (or something)?
I guess what I am trying to understand is what is the generic way to do a load testing using pushy? 

Any advise is appreciated.

**Comment by jchambers** [comment, marked as answer]
> With this context does APNs have any restriction if it observes load coming from one single source and going to one/two device tokens?

According to ["Handling Notification Responses from APNs"](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/handling_notification_responses_from_apns), the server may respond with "TooManyRequests" if "too many requests were made consecutively to the same device token." The precise limit isn't specified.

> Will there be any chance that our app may get blocked (or something) or IPs are blocked (or something)?

I think this is a question for Apple; I don't know for sure.

> I guess what I am trying to understand is what is the generic way to do a load testing using pushy?

You may find Pushy's local [benchmarks](https://github.com/jchambers/pushy/tree/master/benchmark) helpful. Pushy also includes a [mock server](https://pushy-apns.org/apidocs/latest/com/eatthepath/pushy/apns/server/MockApnsServer.html) that might be helpful.

**Reply by inventriz** [reply]
Thanks @jchambers for your prompt reply.

**Reply by inventriz** [reply]
@jchambers - just as a follow up question... do you know how to reach apple to know this - "Will there be any chance that our app may get blocked (or something) or IPs are blocked (or something)?"
I have a developer id but not able to get thru this.

**Reply by jchambers** [reply]
I don't know. You could try the [developer forums](https://developer.apple.com/forums/), but this seems like the kind of thing that Apple generally doesn't disclose because (a) it's subject to change and (b) malicious actors (i.e. spammers) could use the information to try to circumvent whatever limits Apple has in place.
